src/backend/catalog/storage_pending_deletes_redo.c

```diff
@@ -0,0 +1,319 @@
+#include "postgres.h"
+
+#include "access/clog.h"
+#include "access/transam.h"
+#include "catalog/storage_pending_deletes.h"
+#include "miscadmin.h"
+#include "storage/md.h"
+#include "utils/elog.h"
+#include "utils/guc.h"
+#include "utils/hsearch.h"
+
+/*
+ * HTAB entry for pending deletes for the given xid.
+ */
+typedef struct PendingDeleteHtabNode
+{
+	TransactionId xid;
+	List	   *relnode_list;	/* list of RelFileNodePendingDelete */
+}	PendingDeleteHtabNode;
+
+/*
+ * Hash table for pending deletes relfilenodes for a given xid.
+ */
+static HTAB *pendingDeletesRedo = NULL;
+
+/*
+ * This function inserts XLOG_PENDING_DELETE record into WAL. It is called
+ * during checkpoint creation.
+ */
+void
+PdlXLogInsert()
+{
+	Size		size = 0;
+
+	if (IsBootstrapProcessingMode() || !gp_track_pending_delete)
+		return;
+
+	PendingRelXactDeleteArray *arr = PdlXLogShmemDump(&size);
+
+	if (size > 0 && arr != NULL)
+	{
+		XLogRecPtr	rec;
+		XLogRecData rdata;
+
+		rdata.buffer = InvalidBuffer;
+		rdata.data = (char *) arr;
+		rdata.len = size;
+		rdata.next = NULL;
+		rdata.buffer_std = false;
```

Comment:
When someday a field will be added to XLogRecData, then it will be initialized if we write so.
```suggestion
		XLogRecData rdata = {
		    .buffer = InvalidBuffer;
		    .data = (char *) arr;
		    .len = size;
                }
```